frontend/device/android/device.lua

```diff
@@ -109,6 +109,18 @@ function Device:retrieveNetworkInfo()
     end
 end
 
+function Device:update(url, filename)
+    -- download the update file if not present
+    local is_already_downloaded = android.download(url, filename)
```

Comment:
Besides being less enthused with this concept tout court (see below), wouldn't this always return true after the first time?

Reply:
it returns true if the file exists, thats why I think we should add version to downloaded apks.

Reply:
What do you mean "add version to downloaded apks"? Why are you removing it in the first place?

Reply:
The update file at http://ota.koreader.rocks/koreader-android-latest-nightly contains this:

```
koreader-android-arm-linux-androideabi-v2019.02-72-gb48b0d2_2019-02-22.apk
```

This will point the link variable to:
```
http://ota.koreader.rocks/koreader-android-arm-linux-androideabi-v2019.02-72-gb48b0d2_2019-02-22.apk
```

But you're replacing it with a mysteriously named file instead.

Reply:
> What do you mean by "add version to downloaded apks"? Why are you removing it in the first place? 

I didn't remove a thing. It is just that I need to specify a name for saving a url. The file gets downloaded to /sdcard/Downloads with x name. x name is not obtained from url but specified manually. That's why I want to replace the "koreader-latest.apk" whatever name for the apk is inside the link variable.

Reply:
I see. There does seem to be something built into Android that might help: https://developer.android.com/reference/android/webkit/URLUtil.html#guessFileName(java.lang.String,%20java.lang.String,%20java.lang.String)

But as far as I'm concerned your proposed solution is essentially fine. Just change this line to include `ota_package`.
https://github.com/koreader/koreader/blob/260427d226343a9d2d74f5bf448416452ac8f907/frontend/ui/otamanager.lua#L195

With `return ota_version, local_version, link, ota_package` you'll have the filename as the fourth return value.

Reply:
Thanks, done!